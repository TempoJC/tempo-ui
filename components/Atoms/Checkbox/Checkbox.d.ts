export default Checkbox;
/**
 * Base component to define a checkbox.
 */
declare const Checkbox: import('../../../../node_modules/react').ForwardRefExoticComponent<import('../../../../node_modules/react').RefAttributes<any>>;
