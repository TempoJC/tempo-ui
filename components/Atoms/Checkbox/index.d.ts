import { CommonProps, ControlledUncontrolledValueProps } from '../../../types';
import { ReactNode, ChangeEvent, KeyboardEvent, MouseEventHandler, FocusEventHandler, Ref } from '../../../../node_modules/react';
export type CheckboxProps = CommonProps &
  ControlledUncontrolledValueProps<boolean, KeyboardEvent<HTMLInputElement> | ChangeEvent<HTMLInputElement>> & {
    id?: string;
    inputValue?: unknown;
    indeterminate?: boolean;
    label?: ReactNode;
    name?: string;
    disabled?: boolean;
    readOnly?: boolean;
    onClick?: MouseEventHandler<HTMLInputElement>;
    onBlur?: FocusEventHandler<HTMLInputElement>;
    onFocus?: FocusEventHandler<HTMLInputElement>;
    tabIndex?: number;
    ref?: Ref<HTMLInputElement>;
    tabIndexValues?: {
      input?: number;
    };
  };

declare function Checkbox(props: CheckboxProps): JSX.Element;

export type { CheckboxProps };
export default Checkbox;
