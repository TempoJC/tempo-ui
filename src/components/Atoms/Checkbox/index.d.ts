import { CommonProps, ControlledUncontrolledValueProps } from "@/types";
import type { ReactNode, ChangeEvent, KeyboardEvent, MouseEventHandler, FocusEventHandler, Ref } from "react";

export type CheckboxProps = CommonProps &
  ControlledUncontrolledValueProps<boolean, KeyboardEvent<HTMLInputElement> | ChangeEvent<HTMLInputElement>> & {
    id?: string;
    inputValue?: unknown;
    indeterminate?: boolean;
    label?: ReactNode;
    name?: string;
    disabled?: boolean;
    readOnly?: boolean;
    onClick?: MouseEventHandler<HTMLInputElement>;
    onBlur?: FocusEventHandler<HTMLInputElement>;
    onFocus?: FocusEventHandler<HTMLInputElement>;
    tabIndex?: number;
    ref?: Ref<HTMLInputElement>;
    tabIndexValues?: {
      input?: number;
    };
  };

declare function Checkbox(props: CheckboxProps): JSX.Element;

export type { CheckboxProps };
export default Checkbox;
