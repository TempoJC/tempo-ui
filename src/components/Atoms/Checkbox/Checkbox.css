.tempo-checkbox {
  --checkbox-svg-check-light: url("data:image/svg+xml;charset=utf-8,%3Csvg width=%2711%27 height=%278%27 xmlns=%27http://www.w3.org/2000/svg%27 fill=%27%23fff%27%3E%3Cpath fill-rule=%27evenodd%27 clip-rule=%27evenodd%27 d=%27M10.67 1.362 4.047 7.674 0 4.204l.651-.76 3.36 2.882L9.982.638l.69.724Z%27/%3E%3C/svg%3E");
  --checkbox-svg-rect-light: url("data:image/svg+xml;charset=utf-8,%3Csvg width=%278%27 height=%272%27 xmlns=%27http://www.w3.org/2000/svg%27 fill=%27%23fff%27%3E%3Cpath d=%27M0 .5h8v1H0z%27/%3E%3C/svg%3E");
  --checkbox-svg-check-dark: url("data:image/svg+xml;charset=utf-8,%3Csvg width=%2711%27 height=%278%27 xmlns=%27http://www.w3.org/2000/svg%27%3E%3Cpath fill-rule=%27evenodd%27 clip-rule=%27evenodd%27 d=%27M10.67 1.362 4.047 7.674 0 4.204l.651-.76 3.36 2.882L9.982.638l.69.724Z%27/%3E%3C/svg%3E");
  --checkbox-svg-rect-dark: url("data:image/svg+xml;charset=utf-8,%3Csvg width=%278%27 height=%272%27 xmlns=%27http://www.w3.org/2000/svg%27%3E%3Cpath d=%27M0 .5h8v1H0z%27/%3E%3C/svg%3E");

  display: inline-flex;
  align-items: center;
  cursor: pointer;
  position: relative;

  /* create a context for the input layer */
  isolation: isolate;
}

.tempo-checkbox__input {
  opacity: 0;
  position: absolute;

  /*
   * Must render above the rest to make sure it handles
   * the click event. Otherwise, weird bugs happen (see #1073)
   */
  z-index: 1;

  cursor: pointer;
  width: var(--size4);
  height: var(--size4);
  left: 2px;
  top: 2px;
}

.tempo-checkbox__icon {
  box-sizing: border-box;
  background: transparent;
  display: block;
  width: var(--size5);
  height: var(--size5);
  flex-shrink: 0;
  position: relative;
  padding: 2px;
}

.tempo-checkbox__icon:before {
  transition: background-color ease 300ms;
  display: block;
  content: "";
  box-sizing: border-box;
  width: 100%;
  height: 100%;
  background: var(--color-token-light-background-1);
  border: 1px solid var(--color-token-light-border-1);
}

.tempo-checkbox__icon:after {
  display: block;
  content: "";
  box-sizing: border-box;
  width: var(--size4);
  height: var(--size4);
  position: absolute;
  left: 2px;
  top: 2px;
}

.tempo-checkbox--indeterminate {
  & .tempo-checkbox__icon:after {
    background: no-repeat center var(--checkbox-svg-rect-light);
  }
}

.tempo-checkbox--checked {
  & .tempo-checkbox__icon:after {
    background: no-repeat center var(--checkbox-svg-check-light);
  }
}

.tempo-checkbox--indeterminate,
.tempo-checkbox--checked {
  & .tempo-checkbox__icon:before {
    background: var(--color-token-light-semantic-highlight);
    border-color: var(--color-token-light-semantic-highlight);
  }
}

.tempo-checkbox__input:focus + .tempo-checkbox__icon:before {
  border: 1px solid var(--color-token-light-semantic-highlight-2);
}

.tempo-checkbox__input:focus:not(:focus-visible) + .tempo-checkbox__icon:before {
  border: 1px solid var(--color-token-light-border-1);
}

.tempo-checkbox--indeterminate :focus + .tempo-checkbox__icon:before,
.tempo-checkbox--checked :focus + .tempo-checkbox__icon:before {
  background: var(--color-token-light-semantic-highlight-2);
  border-color: var(--color-token-light-semantic-highlight-2);
}

.tempo-checkbox--indeterminate :focus:not(:focus-visible) + .tempo-checkbox__icon:before,
.tempo-checkbox--checked :focus:not(:focus-visible) + .tempo-checkbox__icon:before {
  background: var(--color-token-light-semantic-highlight);
  border-color: var(--color-token-light-semantic-highlight);
}

.tempo-checkbox--disabled.tempo-checkbox--indeterminate :focus:not(:focus-visible) + .tempo-checkbox__icon:before,
.tempo-checkbox--read-only.tempo-checkbox--checked :focus:not(:focus-visible) + .tempo-checkbox__icon:before {
  background: var(--color-token-light-semantic-highlight);
  border-color: var(--color-token-light-semantic-highlight);
}

.tempo-checkbox--read-only {
  cursor: default;
  pointer-events: none;
}

.tempo-checkbox--disabled,
.tempo-checkbox--read-only {
  cursor: default;
  & .tempo-checkbox__icon:before {
    background: var(--color-token-light-background-disabled);
    border-color: var(--color-token-light-border-3);
  }
  &.tempo-checkbox--indeterminate,
  &.tempo-checkbox--checked {
    & .tempo-checkbox__icon:before {
      background: var(--color-token-light-background-3);
      border-color: var(--color-token-light-background-3);
    }
  }
}

.tempo-checkbox--disabled.tempo-checkbox--indeterminate :focus:not(:focus-visible) + .tempo-checkbox__icon:before,
.tempo-checkbox--disabled.tempo-checkbox--checked :focus:not(:focus-visible) + .tempo-checkbox__icon:before {
  background: var(--color-token-light-background-3);
  border-color: var(--color-token-light-background-3);
}

.tempo-checkbox--disabled {
  & .tempo-checkbox__label {
    color: var(--color-token-light-content-disabled);
  }
}

.tempo-checkbox__label {
  box-sizing: border-box;
  margin-left: var(--size2);
  font: var(--typography-body-m-light-font);
  letter-spacing: var(--typography-body-m-light-letter-spacing);
  color: var(--color-token-light-content-1);
}

@media (--viewport-xs-m) {
  .tempo-checkbox__label {
    font: var(--typography-body-l-light-font);
    letter-spacing: var(--typography-body-l-light-letter-spacing);
  }
}

/** Dark **/
.tempo-checkbox--color-dark {
  & .tempo-checkbox__icon:before {
    background: var(--color-token-dark-background-1);
    border: 1px solid var(--color-token-dark-border-1);
  }

  &.tempo-checkbox--indeterminate,
  &.tempo-checkbox--checked {
    & .tempo-checkbox__icon:before {
      background: var(--color-token-dark-semantic-highlight);
      border-color: var(--color-token-dark-semantic-highlight);
    }
  }

  &.tempo-checkbox--disabled,
  &.tempo-checkbox--read-only {
    & .tempo-checkbox__icon:before {
      background: var(--color-token-dark-background-disabled);
      border-color: var(--color-token-dark-border-3);
    }
    &.tempo-checkbox--indeterminate,
    &.tempo-checkbox--checked {
      & .tempo-checkbox__icon:before {
        background: var(--color-token-dark-background-3);
        border-color: var(--color-token-dark-background-3);
      }
    }
  }

  &.tempo-checkbox--disabled {
    & .tempo-checkbox__label {
      color: var(--color-token-dark-content-disabled);
    }
  }

  &.tempo-checkbox--indeterminate {
    & .tempo-checkbox__icon:after {
      background: no-repeat center var(--checkbox-svg-rect-dark);
    }
  }
  &.tempo-checkbox--checked {
    & .tempo-checkbox__icon:after {
      background: no-repeat center var(--checkbox-svg-check-dark);
    }
  }
  & .tempo-checkbox__label {
    color: var(--color-token-dark-content-1);
  }
}

.tempo-checkbox--color-dark .tempo-checkbox__input:focus + .tempo-checkbox__icon:before {
  border: 1px solid var(--color-token-dark-semantic-highlight-2);
}

.tempo-checkbox--color-dark .tempo-checkbox__input:focus:not(:focus-visible) + .tempo-checkbox__icon:before {
  border: 1px solid var(--color-token-dark-border-1);
}

.tempo-checkbox--color-dark.tempo-checkbox--checked :focus + .tempo-checkbox__icon:after {
  background: no-repeat center var(--checkbox-svg-check-light);
}

.tempo-checkbox--color-dark.tempo-checkbox--checked :focus:not(:focus-visible) + .tempo-checkbox__icon:after {
  background: no-repeat center var(--checkbox-svg-check-dark);
}

.tempo-checkbox--color-dark.tempo-checkbox--indeterminate :focus + .tempo-checkbox__icon:after {
  background: no-repeat center var(--checkbox-svg-rect-light);
}

.tempo-checkbox--color-dark.tempo-checkbox--indeterminate :focus:not(:focus-visible) + .tempo-checkbox__icon:after {
  background: no-repeat center var(--checkbox-svg-rect-dark);
}

.tempo-checkbox--color-dark.tempo-checkbox--indeterminate :focus + .tempo-checkbox__icon:before,
.tempo-checkbox--color-dark.tempo-checkbox--checked :focus + .tempo-checkbox__icon:before {
  background: var(--color-token-dark-semantic-highlight-2);
  border-color: var(--color-token-dark-semantic-highlight-2);
}

.tempo-checkbox--color-dark.tempo-checkbox--indeterminate :focus:not(:focus-visible) + .tempo-checkbox__icon:before,
.tempo-checkbox--color-dark.tempo-checkbox--checked :focus:not(:focus-visible) + .tempo-checkbox__icon:before {
  background: var(--color-token-dark-semantic-highlight);
  border-color: var(--color-token-dark-semantic-highlight);
}
