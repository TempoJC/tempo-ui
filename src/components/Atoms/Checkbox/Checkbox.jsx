import { forwardRef } from "react";
import cn from "classnames";
import { useTabIndexValues } from "@/hooks";
import { keystroke } from "@/utils";
import "./Checkbox.css";

/**
 * Base component to define a checkbox.
 */
const Checkbox = forwardRef(
  (
    {
      id,
      name,
      label,
      className,
      inputValue,
      indeterminate = false,
      readOnly = false,
      disabled = false,
      testId,
      color = "light",
      value,
      onChange,
      onClick,
      onFocus,
      onBlur,
      tabIndex,
      inputProps,
    },
    ref,
  ) => {
    const tabIndexValues = useTabIndexValues();
    const handleChange = (value, e) => {
      if (!readOnly && onChange) {
        onChange(value, e);
      }
    };

    const defaultClassName = cn("tempo-checkbox", className, {
      "tempo-checkbox--checked": value,
      "tempo-checkbox--disabled": disabled,
      "tempo-checkbox--read-only": readOnly,
      "tempo-checkbox--indeterminate": !value && indeterminate,
      "tempo-checkbox--color-dark": color === "dark",
    });

    return (
      <label className={defaultClassName} data-testid={testId}>
        <input
          data-testid={testId && `${testId}-input`}
          className="tempo-checkbox__input"
          id={id}
          name={name}
          value={inputValue}
          checked={value}
          disabled={disabled}
          readOnly={readOnly}
          ref={ref}
          onChange={(e) => handleChange(e.target.checked, e)}
          onFocus={onFocus}
          onBlur={onBlur}
          type="checkbox"
          onClick={onClick}
          tabIndex={tabIndexValues.input ?? tabIndex}
          {...(readOnly ? { tabIndex: -1 } : {})}
          onKeyPress={(e) => {
            if (keystroke(e).enter) {
              e.preventDefault();
              handleChange(!e.target.checked, e);
            }
          }}
          {...inputProps}
        />
        <span className="tempo-checkbox__icon" />
        {label && (
          <span className="tempo-checkbox__label" data-testid={testId && `${testId}-label`}>
            {label}
          </span>
        )}
      </label>
    );
  },
);

export default Checkbox;
