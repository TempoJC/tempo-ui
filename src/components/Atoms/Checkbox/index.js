// packages/components-content/checkbox
export { default } from "./Checkbox.jsx";
