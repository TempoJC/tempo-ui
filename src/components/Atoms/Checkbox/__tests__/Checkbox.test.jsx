import { createRef } from "react";
import { userEvent } from "@testing-library/user-event";
import { render, screen, fireEvent } from "@testing-library/react";
import Checkbox from "../Checkbox";

describe("@tempo-ui/Conteng/Checkbox", () => {
  it("renders", () => {
    const { asFragment } = render(<Checkbox />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("adds a name", () => {
    render(<Checkbox name="nameForm" />);
    expect(screen.queryByRole("checkbox")).toHaveAttribute("name", "nameForm");
  });

  it("adds a className", () => {
    const { container } = render(<Checkbox className="test-class" />);

    expect(container.querySelector("label")).toHaveClass("test-class");
  });

  it("adds an inputValue", () => {
    render(<Checkbox inputValue="test-value" />);
    expect(screen.queryByRole("checkbox")).toHaveAttribute("value", "test-value");
  });

  it("adds disabled", async () => {
    const onClick = jest.fn();

    const { container } = render(<Checkbox onClick={onClick} disabled />);
    const checkbox = screen.queryByRole("checkbox");

    expect(checkbox).toHaveAttribute("disabled");
    expect(container.querySelector("label")).toHaveClass("tempo-checkbox--disabled");

    await userEvent.click(checkbox);

    expect(onClick).not.toHaveBeenCalled();
  });

  it("adds readOnly", async () => {
    const onChange = jest.fn();

    const { container } = render(<Checkbox onChange={onChange} readOnly />);
    const checkbox = screen.queryByRole("checkbox");
    expect(checkbox).toHaveAttribute("readOnly");
    expect(container.querySelector("label")).toHaveClass("tempo-checkbox--read-only");

    await userEvent.click(checkbox);

    expect(onChange).not.toHaveBeenCalled();
  });

  it("adds events", async () => {
    const onChange = jest.fn();
    const onFocus = jest.fn();
    const onBlur = jest.fn();
    render(<Checkbox value={false} onChange={onChange} onFocus={onFocus} onBlur={onBlur} testId="checkbox" />);
    const checkbox = screen.queryByRole("checkbox");

    await userEvent.click(checkbox);
    expect(onChange).toHaveBeenCalled();

    fireEvent.focus(checkbox);
    expect(onFocus).toHaveBeenCalled();

    fireEvent.blur(checkbox);
    expect(onBlur).toHaveBeenCalled();
  });

  it("adds inputRef", () => {
    let inputRef = createRef();
    render(<Checkbox ref={inputRef} inputValue={"test-value"} />);
    expect(inputRef.current.value).toBe("test-value");
  });

  it("adds an id", () => {
    render(<Checkbox id="testId" />);
    expect(screen.queryByRole("checkbox")).toHaveAttribute("id", "testId");
  });

  it("adds color", () => {
    const { container } = render(<Checkbox color="dark" />);
    expect(container.querySelector("label")).toHaveClass("tempo-checkbox--color-dark");
  });

  it("adds indeterminate", () => {
    const { container, rerender } = render(<Checkbox indeterminate value={false} />);
    const label = container.querySelector("label");
    expect(label).toHaveClass("tempo-checkbox--indeterminate");
    expect(label).not.toHaveClass("tempo-checkbox--checked");

    rerender(<Checkbox indeterminate value={true} />);
    expect(label).not.toHaveClass("tempo-checkbox--indeterminate");
    expect(label).toHaveClass("tempo-checkbox--checked");
  });

  it("changes value", () => {
    const { container, rerender } = render(<Checkbox value={true} />);
    const checkbox = screen.queryByRole("checkbox");
    const label = container.querySelector("label");

    expect(checkbox).toBeChecked();
    expect(label).toHaveClass("tempo-checkbox--checked");

    rerender(<Checkbox value={false} />);

    expect(checkbox).not.toBeChecked();
    expect(label).not.toHaveClass("tempo-checkbox--checked");
  });

  it("forwards tabIndex", () => {
    render(<Checkbox value={true} tabIndex={2} />);
    expect(screen.queryByRole("checkbox")).toHaveAttribute("tabIndex", "2");
  });

  describe("testId prop", () => {
    it("renders the Checkbox with the specified data-testid attribute", () => {
      const { getByTestId } = render(<Checkbox testId="checkbox" />);

      const element = getByTestId("checkbox");

      expect(element).toBeInTheDocument();
      expect(element).toHaveAttribute("data-testid", "checkbox");
    });
  });
});
