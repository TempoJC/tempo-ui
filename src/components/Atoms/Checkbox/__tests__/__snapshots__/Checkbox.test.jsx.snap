// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`@tempo-ui/Conteng/Checkbox renders 1`] = `
<DocumentFragment>
  <label
    class="tempo-checkbox"
  >
    <input
      class="tempo-checkbox__input"
      type="checkbox"
    />
    <span
      class="tempo-checkbox__icon"
    />
  </label>
</DocumentFragment>
`;
